haystack/database/faiss.py

```diff
@@ -0,0 +1,173 @@
+
+    It implements the FAISS library(https://github.com/facebookresearch/faiss)
+    to perform similarity search on vectors.
+
+    The document text and meta-data(for filtering) is stored using the SQLDocumentStore, while
+    the vector embeddings are indexed in a FAISS Index.
+
+    """
+
+    def __init__(
+        self,
+        sql_url: str = "sqlite:///",
+        index_factory: str = "HNSW4",
+        index_buffer_size: int = 10_000,
+        vector_size: int = 768,
+    ):
+        """
+        :param sql_url: SQL connection URL for database. It defaults to local file based SQLite DB. For large scale
+                        deployment, Postgres is recommended.
+        :param index_factory: FAISS provides a function to build composite index based on comma separated list of
+                              components. The FAISS documentation has guidelines for choosing
+                              an index: https://github.com/facebookresearch/faiss/wiki/Guidelines-to-choose-an-index.
+                              More details on index_factory: https://github.com/facebookresearch/faiss/wiki/The-index-factory
+
+                              The default index here is fast with good accuracy at the cost of higher memory usage. It
+                              does not require any training.
+        :param index_buffer_size: When working with large dataset, the indexing process(FAISS + SQL) can be buffered in
+                                  smaller chunks to reduce memory footprint.
+        :param vector_size: the embedding vector size.
+        """
+        self.vector_size = vector_size
+        self.faiss_index: IndexHNSWFlat = self._create_new_index(index_factory=index_factory, vector_size=vector_size)
+        self.index_factory = index_factory
+
+        self.index_buffer_size = index_buffer_size
+        self.phi = 0
+        super().__init__(url=sql_url)
+
+    def _create_new_index(self, index_factory: str, vector_size: int):
+        index = faiss.index_factory(vector_size + 1, index_factory)
+        return index
+
+    def write_documents(self, documents: Union[List[dict], List[Document]], index: Optional[str] = None):
+        if self.phi > 0:
+            raise Exception("Addition of more data in an existing index is not supported.")
+
+        index = index or self.index
+        document_objects = [Document.from_dict(d) if isinstance(d, dict) else d for d in documents]
+
+        add_vectors = False if document_objects[0].embedding is None else True
+
+        if add_vectors:
+            phi = self._get_phi(document_objects)
+            self.phi = 0
+
+        for i in range(0, len(document_objects), self.index_buffer_size):
+            if add_vectors:
+                embeddings = [doc.embedding for doc in document_objects[i : i + self.index_buffer_size]]
+                hnsw_vectors = self._get_hnsw_vectors(embeddings=embeddings, phi=phi)
```

Comment:
isnt self._get_hnsw_vectors dependent on self.index_factory and should be switched?

Reply:
That's a good point. Should we remove the `self.index_factory` parameter for now to make it simpler as we currently only implement the HNSW index?